src/credentials/sso/ssoAccessTokenProvider.ts

```diff
@@ -0,0 +1,148 @@
+        )
+
+        // The retry interval converted to milliseconds
+        let retryInterval = authorization.interval * 1000
+
+        const createTokenParams = {
+            clientId: registration.clientId,
+            clientSecret: registration.clientSecret,
+            grantType: GRANT_TYPE,
+            deviceCode: authorization.deviceCode,
+        }
+
+        while (true) {
+            try {
+                const tokenResponse = await this.ssoOidcClient.createToken(createTokenParams).promise()
+                const accessToken: AccessToken = {
+                    startUrl: this.ssoUrl,
+                    region: this.ssoRegion,
+                    accessToken: tokenResponse.accessToken!,
+                    expiresAt: new Date(Date.now() + tokenResponse.expiresIn! * 1000).toISOString(),
+                }
+                return accessToken
+            } catch (err) {
+                if (err.code === 'SlowDownException') {
+                    retryInterval += 5000
+                } else if (err.code === 'AuthorizationPendingException') {
+                    if (Date.now() + retryInterval > deviceCodeExpiration) {
+                        throw Error(
+                            `Device code has expired while polling for SSO token, login flow must be re-initiated.`
+                        )
+                    }
+                    // else, wait the interval and try again
+                } else if (err.code === 'ExpiredTokenException') {
+                    throw Error(`Device code has expired while polling for SSO token, login flow must be re-initiated.`)
+                }
+            }
+            setTimeout(() => {}, retryInterval)
+        }
+    }
+
+    private async authorizeClient(registration: ClientRegistration): Promise<Authorization> {
+        const authorizationParams = {
+            clientId: registration.clientId,
+            clientSecret: registration.clientSecret,
+            startUrl: this.ssoUrl,
+        }
+        try {
+            const authorizationResponse = await this.ssoOidcClient
+                .startDeviceAuthorization(authorizationParams)
+                .promise()
+            const authorization: Authorization = {
+                deviceCode: authorizationResponse.deviceCode!,
+                userCode: authorizationResponse.userCode!,
+                verificationUri: authorizationResponse.verificationUri!,
+                verificationUriComplete: authorizationResponse.verificationUriComplete!,
+                expiresIn: authorizationResponse.expiresIn!,
+                interval: authorizationResponse.interval!,
+            }
+
+            const signInInstructionMessage = `You have chosen an AWS Single Sign-On profile that requires authorization. `
+            vscode.window.showInformationMessage(signInInstructionMessage, { modal: true })
+            vscode.env.openExternal(vscode.Uri.parse(authorization.verificationUriComplete))
```

Comment:
We may want to split this out to a callback. It isn't the concern of the code that is acquiring a token, rather somewhere up the call chain would be driving this behavior. 

This also sets you up to be more compatible with an SDK implementation.